### PR TITLE
Fix the dll load flag for Linux

### PR DESCRIPTION
**Pull request by mcfatealan**

### CODE REVIEW

File: src/core/core/library_utils.cpp [hunk 1]
```diff
@@ -71,7 +71,7 @@ namespace dsn {
             }            
 # elif defined(__linux__) || defined(__FreeBSD__) || defined(__APPLE__)
             module_name = "lib" + module_name + ".so";
-            auto hmod = dlopen(module_name.c_str(), RTLD_LAZY);
+            auto hmod = dlopen(module_name.c_str(), RTLD_LAZY|RTLD_GLOBAL);
```
Comment (qinzuoyan) [inline src/core/core/library_utils.cpp:74]: Why global here? worry about polluting the global symbol table.

Comment (mcfatealan) [inline src/core/core/library_utils.cpp:74]: _Hi zuoyan, thanks for pointing it out. My bad, I should explain the context more clearly in the commit history._
The reason is that when I was trying to deploy rDSN.monitor on Linux, I met **ImportError** such as:

``` python
undefined symbol: _Py_ZeroStruct, PyFloat_Type
```

And the load link is like:
**simple_kv(executable)** --dynamically-load-> **libdev.python_helper.so** --load-> **python interpreter**
To solve this issue, I thought using RTLD_GLOBAL would make symbols available for libraries loaded afterwards.

**Yes**, your concern is right. Another alternative is to link Python lib with executables, to which I hold my opinion. I would consult Zhenyu tomorrow to see what he would suggest.

Comment (imzhenyu) [inline src/core/core/library_utils.cpp:74]: it seems the missing symbol is in Python, which we don't have control. According to http://stackoverflow.com/questions/480617/receive-undefined-symbol-error-when-loading-libary-with-dlopen, we probably should use global. Not 100 percent sure though:-) The dynamic link process of Linux just sucks~

Comment (mcfatealan) [inline src/core/core/library_utils.cpp:74]: Another related question is http://stackoverflow.com/questions/8302810/undefined-symbol-in-c-when-loading-a-python-shared-library, which is very similar to our situation. The "GLOBAL" solution happens to be the highest-voted answer.

Comment (imzhenyu) [inline src/core/core/library_utils.cpp:74]: @qinzuoyan let's stick to the GLOBAL solution for the time being. You concern is definitely correct, and we can fix this later when we have a better solution. Right now, the solution we see is to link python with our main program, which is not acceptable.

**Comment by imzhenyu** [comment]
the above ci failure is caused by ZK test cases.
[0;32m[ RUN      ] [mmeta_state_service_zookeeper.basics
./run.sh: line 4:  9658 Segmentation fault      (core dumped) ./dsn.tests config-test.ini

I'm merging this pr now.

**Comment by qinzuoyan** [comment]
I see, I also meet the same problem in another project before, and add RTLD_GLOBAL is an acceptable resolution. And more, my suggestion is to add a config param such as **_"runtime_load_global = true"**_ (along with "dmodule") to make it configurable, and the default value would be false for normal cases.

**Comment by qinzuoyan** [comment]
Good news, now the travis ci can print gdb backtrace when coredump, which will help us to fix coredump problems which may only occurs on travis.

@imzhenyu , as your comment above, the ci failure is caused by zk test cases, and now it is fixed in https://github.com/qinzuoyan/rDSN/commit/ec2da900b1726c991219f9b03ac5e961ca6bc80b. As you see, it's a common problem caused by improper use of **_auto**_ keyword that we have meet before. So we should be careful when using auto, and I also did more refactors to avoid risk in https://github.com/qinzuoyan/rDSN/commit/fa98ab54574d2a8b20ce74e3f8fb31d791b7b0d2.

**Comment by imzhenyu** [comment]
nice, @qinzuoyan. The auto keyword problems raised before, as far as I know, are caused by the improper return type of create_late_task. You may also fix that too if not too much trouble for you:) Waiting for your pull requests then.

**Pull request merged**